ghcide/src/Development/IDE/Core/Shake.hs

```diff
@@ -573,11 +576,22 @@ recordDirtyKeys
   -> k
   -> [NormalizedFilePath]
   -> STM (IO ())
-recordDirtyKeys ShakeExtras{dirtyKeys} key file = do
+recordDirtyKeys ShakeExtras{dirtyKeys, runningKeys} key file = do
+    modifyTVar' runningKeys $ \x -> foldl' (flip deleteKeySet) x (toKey key <$> file)
     modifyTVar' dirtyKeys $ \x -> foldl' (flip insertKeySet) x (toKey key <$> file)
     return $ withEventTrace "recordDirtyKeys" $ \addEvent -> do
         addEvent (fromString $ unlines $ "dirty " <> show key : map fromNormalizedFilePath file)
 
+recordDirtyKeySet
+  :: ShakeExtras
+  -> [Key]
+  -> STM (IO ())
+recordDirtyKeySet ShakeExtras{dirtyKeys, runningKeys} keys = do
+    modifyTVar' runningKeys $ \x -> foldl' (flip deleteKeySet) x keys
```

Comment:
I think this is a little counter-intuitive. It means that we will remove a key from the running set _while it is in fact running_. But I think maybe we just need to rename the "running keys" set. It's something like "running keys which will produce non-dirty results". With that interpretation it is fine to remove the key here, since we are denying that it will produce a non-dirty result when it finishes.

This also all definitely needs a Note to explain what's going on!

Reply:
Yep, maybe rename to something like `runningCleanlyKeys` to indicate that it have not been polluted

Reply:
done